runtime/reachable.c

```diff
@@ -0,0 +1,413 @@
+    is32Bit = fh->e_ident[EI_CLASS] == 1;
+    if (is32Bit ^ (sizeof(ELF_Uint) == 4)) {
+      rv = "32/64-bit mismatch in executable file";
+      break;
+    }
+
+    p = (char *)&my_endian;
+    *p = 1;
+    isLittleEndian = fh->e_ident[EI_DATA] == 1;
+    if (isLittleEndian ^ (my_endian == 1)) {
+      rv = "endianness mismatch in executable file";
+      break;
+    }
+
+    if (fh->e_ident[EI_VERSION] != 1 || fh->e_version != 1) {
+      rv = "not ELF V1";
+      break;
+    }
+
+    section_hdrs = (void *)buf + fh->e_shoff;
+    shrstrtab_hdr = &section_hdrs[fh->e_shstrndx];
+    elf_info->shrstrtab = buf + shrstrtab_hdr->sh_offset;
+
+    sh = find_section(elf_info, ".strtab");
+    if (sh == NULL) {
+      rv = "Missing .strtab section";
+      break;
+    }
+    elf_info->strtab = buf + sh->sh_offset;
+    
+    sh = find_section(elf_info, ".symtab");
+    if (sh == NULL) {
+      rv = "Missing .symtab section";
+      break;
+    }
+    elf_info->symtab = (Elf_Sym *)(buf + sh->sh_offset);
+    symtab_entsize = sh->sh_entsize;
+    elf_info->symtab_num_ents = sh->sh_size / symtab_entsize;
+
+    if (symtab_entsize != sizeof (Elf_Sym)) {
+      rv = "Symtab entry size mismatch";
+      break;
+    }
+
+    reloc_sym = caml_reloc_sym;  // avoid "unused variable" warning
+    reloc_sym = caml_get_elf_symbol_by_name(elf_info, "caml_reloc_sym");
+    if (reloc_sym == -1) {
+      rv = "Can't find 'caml_reloc_sym' symbol";
+      break;
+    }
+    elf_info->relocation = ((ELF_Uint)&caml_reloc_sym) - reloc_sym;
+  } while (0);
+
+  if (exe_file != NULL)
+    fclose(exe_file);
+  if (strcmp(rv, "") != 0 && buf != NULL) {
+    elf_info->buf = NULL;
+    free(buf);
+  }
+  return rv;
+}
+
+char *caml_get_elf_symbol_by_address(Elf_info *elf_info, value *addr) {
+  ELF_Uint sym_unrelocated;
+  Elf_Sym *sym_ent;
+  
+  if (elf_info->buf == NULL) {
+    printf("call elf_info_setup before calling caml_get_elf_symbol_by_address\n");
+    return "???";
+  }
+
+  sym_unrelocated = (ELF_Uint)addr - elf_info->relocation;
+  for (int i = 0; i < elf_info->symtab_num_ents; i++) {
+    sym_ent = &elf_info->symtab[i];
+    if (sym_ent->st_value == sym_unrelocated)
+      return strdup(elf_info->strtab + sym_ent->st_name);
+  }
+
+  return "???";  // not found
+}
+
+static struct {
+  ELF_Uint min;          // minimum number of reachable words
+  char *header_string;  // the header string (per section)
+  int printed_header;   // boolean; 1 = header has been printed
+  int item_num;         // simple counter but not very meaningful
+} print_roots;
+
+// callback routine
+void caml_root_report (value v, value *p /* not used */) {
+  value reachable = Long_val(caml_obj_reachable_words(v));
+  if (reachable >= print_roots.min) {
+    if (!print_roots.printed_header) {
+      printf("\n%s:\n", print_roots.header_string);
+      print_roots.printed_header = 1;
+    }
+    printf("%10" ELF_INT_PRINTF_FORMAT "d item  %3d\n",
+        reachable, print_roots.item_num);
```

Comment:
This worked when I compiled it in 32-bit mode using 4.10, but it breaks on `trunk` in 32-bit mode with the message:

```
reachable.c: In function ‘caml_root_report’:
reachable.c:212:12: error: format ‘%d’ expects argument of type ‘int’, but argument 2 has type ‘value {aka long int}’ [-Werror=format=]
     printf("%10" ELF_INT_PRINTF_FORMAT "d item  %3d\n",
            ^~~~~
reachable.c:212:41: note: format string is defined here
     printf("%10" ELF_INT_PRINTF_FORMAT "d item  %3d\n",
             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~^
             %10" ELF_INT_PRINTF_FORMAT "ld
```